cpp/src/strings/copying/concatenate.cu

```diff
@@ -60,39 +66,89 @@ struct chars_size_transform {
   }
 };
 
+template <typename ColumnDeviceView>
+auto create_contiguous_device_views(std::vector<column_view> const& source_view,
```

Comment:
similar logic exists for `table_device_view`. Can these things be consolidated?

Reply:
`table_device_view` takes `table_view` which requires all columns to have same number of rows. 
for  `cudf::concatenate`, each column could have different number of rows. So, `table_device_view` can't be reused.

Reply:
A new type 'list_of_column_views` is added (without that restriction) and the same logic is reused in `list_of_device_column_views` (with begin, end as host function too).

Reply:
> table_device_view takes table_view which requires all columns to have same number of rows.

I'm aware. My point is that the logic inside of `table_device_view` for packing into a buffer is not dependent on having the same number of rows. This common functionality should be consolidated.

Reply:
A similiar problem came up in https://github.com/rapidsai/cudf/pull/6615, so I wrote some code that should be able to be used in this PR as well: https://godbolt.org/z/bssYW1

Reply:
aliasing is not a problem here. This code creates `column_device_view` with placement new operator for all columns.

Reply:
You're doing placement new into an aliased pointer into a single buffer.

Reply:
Here's how I would implement this using `alias_pointers`.

```
std::vector<ColumnDevicView*> host_pointers(source_view.size());
std::vector<std:size_t> sizes(source_view.size());
std::transform(source_view.begin(), source_view.end(), sizes.begin(), [](auto col){ return ColumnDeviceView::extent(col); });

auto bytes = alias_pointers(nullptr, alignof(ColumnDeviceView), host_pointers, sizes);
std::vector<char> host_buffer(bytes);
alias_pointers(host_buffer.data(), alignof(ColumnDeviceView), host_pointers, sizes);

rmm::device_buffer device_buffer(bytes);
std::vector<ColumnDeviceView*> device_pointers(source_view.size());
alias_pointers(device_buffer.data(), alignof(ColumnDeviceView), device_pointers, sizes);

for(int i = 0; i < source_view.size(); ++i){
   auto p = host_pointers[i];
   new (p) ColumnDeviceView(source_view[i], device_pointers[i]); // this logic may not be 100% correct, but you get the general idea
}

cudaMemcpyAsync(device_pointers.data(), host_pointers.data(), bytes, cudaMemcpyDefault, stream);
```

Reply:
This aligns the pointers for the `source_view` children but not the pointers of the children's children. 
The bytes calculation needs to be moved into `extent()` and the `alias_pointers` logic copied into the `(mutable)_column_device_view` ctors.

Reply:
Since we are copying only `column_device_view` only into `device_buffer.data()`, 
we need to align only the first pointer. right?

Rest all should be automatically aligned because
`sizeof(column_device_view)` should be already be multiples of `alignof(column_device_view)`.

Reply:
@jrhemstad 

Let's say there is a column hierarchy like this.
`C1, C2{C21,C22}, C3{C31,C32{C321, C322}, C33{C333}}` 
braces `{}` indicate child columns.

Existing code will create memory layout of ColumnDeviceView like this.
`C1, C2, C3, C21, C22, C31, C32, C33, C321, C322, C333`
This layout also enables `d_children` to be pointer `column_device_view*`

The proposed code https://github.com/rapidsai/cudf/pull/6605#discussion_r514312391 will create a layout like this.
`C1, C2{C21,C22}, C3{C31,C32{C321, C322}, C33{C333}}`
This will not allow `d_children` to be pointer `column_device_view*`. May be vector of pointers.

This logic https://github.com/rapidsai/cudf/pull/6605#discussion_r514312391 could be modified to follow existing memory layout with alignment.
But again, only alignment needed is for first column_device_view `C1`, since all the types are just `column_device_view`.

Reply:
> Since we are copying only column_device_view only into device_buffer.data(),
we need to align only the first pointer. right?

Yes, I believe this is technically correct. Though it would be safer/nicer if `extent` and the `column_device_view` ctors were updated to use logic like `alias_pointers` just to ensure the alignment is correct.

> This will not allow d_children to be pointer column_device_view*. May be vector of pointers.

I don't understand. So long as a column's children at level `i` in the hierarchy are contiguous in memory, why does it matter what the layout is?

Reply:
d_children is a simple pointer now. This expects all children device_view at that level `i` to be **consecutive**. 
https://github.com/rapidsai/cudf/blob/dd3eb62847b763a8579c62d2e7444e4f774e75d2/cpp/include/cudf/column/column_device_view.cuh#L416

with new layout, then `d_children[0]` and `d_children[1]` will not be consecutive. children device_view at that level `i`, will not be consecutive. This layout forces `d_children` to be a vector of pointers.

Reply:
The logic I described in https://github.com/rapidsai/cudf/pull/6605#discussion_r514312391 doesn't impact the layout of the children. The `alias_pointers` logic just sets up where the parent pointers alias in the buffer, i.e., the position of `C1, C2, C3`. 

The layout of the children would be handled here in the `ColumnDeviceView` constructor:  

```
new (p) ColumnDeviceView(source_view[i], device_pointers[i]); // this logic may not be 100% correct, but you get the general idea
```

I haven't worked through it, but I would think you could implement this constructor similarly using `alias_pointers` with appropriate logic to maintain the desired layout. If not, at least `alias_pointers` sets up the correct position for the parent column pointers.

Reply:
Okay, I see. I'm confident you should be able to work out using an abstraction something like the `alias_pointers` I described. This kind of raw pointer manipulation is fragile and time consuming to read. Furthermore, the fact that the existing logic is nearly identical to [`child_columns_to_device_array`](https://github.com/rapidsai/cudf/blob/24dcc8e3a323ab865a57cfb04a455a467199f37b/cpp/src/column/column_device_view.cu#L65-L89) should set off your DRY alarm bells that an abstraction is necessary.